petdeface/pet.py

```diff
@@ -1,42 +1,39 @@
-def create_weighted_average_pet(pet_file, bids_dir):
+def create_weighted_average_pet(pet_file: str, bids_dir: str) -> str:
+    import os
+    from pathlib import Path
 
-    import json
-    from niworkflows.interfaces.bids import ReadSidecarJSON
     import nibabel as nib
     import numpy as np
-    import os
-    from pathlib import Path
+    from niworkflows.interfaces.bids import ReadSidecarJSON
 
     """
     Create a time-weighted average of dynamic PET data using mid-frames
-    
-    Arguments
-    ---------
-    pet_file: string
-        path to input dynamic PET volume
-    bids_dir: string
-        path to BIDS directory containing the PET file
-    """     
-      
-    img = nib.load(pet_file)        
-    data = img.get_fdata()
 
-    meta = ReadSidecarJSON(in_file = pet_file, 
-                           bids_dir = bids_dir, 
-                           bids_validate = False).run()
+    Args:
+        pet_file: path to input dynamic PET volume
+        bids_dir: path to BIDS directory containing the PET file
+
+    Returns:
+        path to time-weighted average of dynamic PET
+    """
+
+    img = nib.load(pet_file)
+    data = img.get_fdata()
 
-    frames_start = np.array(meta.outputs.out_dict['FrameTimesStart'])
-    frames_duration = np.array(meta.outputs.out_dict['FrameDuration'])
+    meta = ReadSidecarJSON(
+        in_file=pet_file, bids_dir=bids_dir, bids_validate=False
+    ).run()
 
-    frames = range(data.shape[-1])
+    frames_start = np.array(meta.outputs.out_dict["FrameTimesStart"])
+    frames_duration = np.array(meta.outputs.out_dict["FrameDuration"])
 
     new_pth = os.getcwd()
 
-    mid_frames = frames_start + frames_duration/2
-    wavg = np.trapz(data[..., frames], dx=np.diff(mid_frames[frames]), axis=3)/np.sum(mid_frames)
+    mid_frames = frames_start + frames_duration / 2
+    wavg = np.trapz(data, x=mid_frames) / (mid_frames[-1] - mid_frames[0])
```

Comment:
What happened here? @mnoergaard

Reply:
According to [numpy.trapz](https://numpy.org/doc/stable/reference/generated/numpy.trapz.html) documentation, `dx` should be a scalar, and I'm not sure what was happening with `dx` being set to an array. Instead, we can specify the location of the data points using the `x` argument, which accepts arrays. The indexing on `data` was redundant, and so was the `axis=3` argument to `np.trapz`, since the default is `axis=-1`. Denominator should be the range of the TAC, not the sum of mid frame time points.

Reply:
Please note that this is a time weighted average (weighted by the frame length) i.e. $W\ =\ \frac{\sum_{i=1}^nw_i\cdot X_i}{\sum_{i=1}^nw_i}$, which matches the formula.

Reply:
You know what, I'm just going to let you two determine which is correct as `np.sum(midframes)` looks very different to me from`midframes[-1] - midframes[0]`. I could walk through this with the debugger, but let's be realistic here.

Reply:
Okay, ah, ah, laying out $w_i$ made this much more clear, thank you.

Reply:
I stand corrected - thank you, Murat! I had originally re-used from here: https://github.com/openneuropet/PET_pipelines/blob/main/pyPetSurfer/pet_surfer.py

But going back and looking at my previous matlab implementation I have this:

function create_tac_wavg(tacpath,time,out)

tac=MRIread(tacpath);
tac.vol=trapz(time,tac.vol,4)/(time(end)-time(1));
MRIwrite(tac,out);